bluecellulab/stimulus/factory.py

```diff
@@ -103,152 +104,286 @@ def current(self) -> np.ndarray:
+                amplitude_start=0.0,
+                amplitude_end=amplitude,
+            )
+            + Empty(dt, duration=post_delay)
+        )
+
+    @classmethod
+    def threshold_based(
+        cls,
+        dt: float,
+        pre_delay: float,
+        duration: float,
+        post_delay: float,
+        threshold_current: float,
+        threshold_percentage: float,
+    ) -> CombinedStimulus:
+        """Creates a Ramp stimulus with respect to the threshold current.
+
+        Args:
+
+            dt: The time step of the stimulus.
+            pre_delay: The delay before the start of the ramp.
+            duration: The duration of the ramp.
+            post_delay: The time to wait after the end of the ramp.
+            threshold_current: The threshold current of the Cell.
+            threshold_percentage: Percentage of desired threshold_current amplification.
+        """
+        amplitude = threshold_current * threshold_percentage / 100
+        res = cls.amplitude_based(
+            dt,
+            pre_delay=pre_delay,
+            duration=duration,
+            post_delay=post_delay,
+            amplitude=amplitude,
+        )
+        return res
+
+
 class StimulusFactory:
     def __init__(self, dt: float):
         self.dt = dt
 
-    def step(self, start: float, end: float, amplitude: float) -> Stimulus:
-        return Step(self.dt, start, end, amplitude)
+    def step(
+        self, pre_delay: float, duration: float, post_delay: float, amplitude: float
+    ) -> Stimulus:
+        return Step.amplitude_based(
+            self.dt,
+            pre_delay=pre_delay,
+            duration=duration,
+            post_delay=post_delay,
+            amplitude=amplitude,
+        )
 
     def ramp(
-        self, start: float, end: float, amplitude_start: float, amplitude_end: float
+        self, pre_delay: float, duration: float, post_delay: float, amplitude: float
     ) -> Stimulus:
-        return Ramp(self.dt, start, end, amplitude_start, amplitude_end)
+        return Ramp.amplitude_based(
+            self.dt,
+            pre_delay=pre_delay,
+            duration=duration,
+            post_delay=post_delay,
+            amplitude=amplitude,
+        )
 
     def ap_waveform(
-        self,
-        threshold_current: float,
-        threshold_percentage: float = 220.0,
-        start: float = 250.0,
-        end: float = 300.0,
-        post_wait: float = 250.0,
+        self, threshold_current: float, threshold_percentage: float = 220.0
     ) -> Stimulus:
         """Returns the APWaveform Stimulus object, a type of Step stimulus.
 
         Args:
             threshold_current: The threshold current of the Cell.
             threshold_percentage: Percentage of desired threshold_current amplification.
-            start: The start time of the step.
-            end: The end time of the step.
-            post_wait: The time to wait after the end of the step.
         """
+        pre_delay = 250.0
```

Comment:
why have pre_delay, duration, and post_delay been hardcoded for the subsequent protocols?

Reply:
Ah yes these 4 protocols are Step protocols that are not modifiable - they have a standard shape. The Step protocol implementation (https://github.com/BlueBrain/BlueCelluLab/pull/164/files/cad47212338bc92d1f33aeb60cae9cbe6ff18f57#diff-6377ba9e4ef26bdb2ee7e7eee3a78082a1e5898940b12d8e4b7e33203c11b50bR279)  can be modified to represent all of them and all of their possible modifications.

Reply:
ha I see, thanks for the explanation